### ISSUE TITLE
How to create a message for an error

### ISSUE THREAD

**Issue by tyler-dunkel** [AdamBrodzinski/validate-form#2]
Hi, 
how would I be able to display a message re-actively based upon when the class "has-error" is added?
Would I use a helper function? and how would I get the helper function to rerun whenever the class is added? 
Thanks

**Comment by AdamBrodzinski** [comment]
Hi @tdmoneybanks 

I'm not exactly sure how to reactively hook into it, but I think the crux of the issue might be the documentation as the package should insert the message for you when the user blurs the input. I compared the docs to my prod code and it looks like you'll need to add a `<span class='err-msg'>  </span>` to the input and wrap it with a `div` so the selector can find it. This is clearly my fault for not including it the docs :smile: 

If the user blurs the input and it's invalid, it should fill in a default message (required or invalid email, etc..). This can be styled accordingly. Eventually I want to make custom messages that override the default ones, and also have some sort of optional callback to render a message on the fly.

Here's a working input block that displays a 'required field` error beside the input whenever the user blurs out of the input:

``` html

  <form id='form-signup' class='validate' novalidate>

      <div class='form-group'>
        <label for='lname'>Last Name</label>
        <span for='lname' class='err-msg'></span>
        <input type='text' name='lname' class='form-control'
          data-required data-onblur>
      </div>     

  </form>
```

**Comment by tyler-dunkel** [comment]
Ok awesome. I had no clue and was planning on adding in some jquery to input the message on the blur event you defined. Do you have any idea on how you would like to implement the custom messages? The main reason I picked this package was because its extremely light weight as I only have 2-3 forms on the webapp FYI and so I think most would like it to say pretty bare-bones. 
Maybe just a data-message attribute that you can define the message you would like for each input?

**Comment by AdamBrodzinski** [comment]
Yep I think you're on the right track. Perhaps download as a local package, and modify it to do something like this?

https://github.com/AdamBrodzinski/validate-form/blob/master/validate-form.js#L194

``` javascript
 _addInputErrorMessage: function(defMessage) {
     var customMsg = this.$el.attr('data-msg');
     var errMsg = (customMsg) ? customMsg : defMessage;

     this.$el.siblings('.err-msg').text(customMsg);
},                    
```

Basically check to see if there is a custom message, if so use it else use the default message. Also if you're doing light form work, checkout my parse-form package... makes it easy to extract out the values.

**Comment by tyler-dunkel** [comment]
Ok. well I showed the client the default messages and he said they will work fine but if I have time ill add it and submit a PR! thanks for the help btw and ill check out your other package but like I said the site has 2-3 forms MAX so not too much is needed ;p

**Comment by AdamBrodzinski** [comment]
Awesome glad to help! I've added the readme fix to my TODOs so that should be resolved soon (i'll leave this open until it's fixed). Also i'm going to be adding some more validations soon as I have a large form project coming up :smiley:

**Comment by tyler-dunkel** [comment]
Awesome! as long as I dont have to do regexs myself I'm happy :)

**Comment by tyler-dunkel** [comment]
Hey quick question! the data-onblur event is not firing for my inputs. here is a gist with my code
https://gist.github.com/tdmoneybanks/9986414fa7735506170a.
Any idea why it would not fire?

**Comment by AdamBrodzinski** [comment]
Hmm very odd. It sounds like the event listeners are not getting setup.

One thing to check is the browser console to see if the layout not getting set and throwing a console.error. If your layout is named 'layout' it will default to that without extra configuration, otherwise it will need to be passed in like below. Speaking of, is the debug spewing out logs if you enable it?

```
ValidateForm.config({
  debug: true   // dumps validations
  rootLayout: 'myLayout'  // only if you need to
});

```

Another thing to debug is to setup your own listener with the same selectors to see if that will fire:

``` javascript
  Template.yourTemplate.events({

    'blur form.validate [data-onblur]': function(e) {
      console.log("validating element", e.target);
    },

    // hide status when user is fixing an error
    'focus .validate input, focus .validate textarea': function(e) {
      console.log('clearing the element');
    }
  });
```

And finally if all else fails (actually trying this first might be best lol), manually pass the input into the validate function:

```
var isValid = ValidateForm.validateInput( $('.yourinput') );
// isValid == true or false
```

Hope this helps! I'm feeling under the weather today but if you're still having issues i'll paste the form in the gist into a meteor project.

**Comment by tyler-dunkel** [comment]
Ha, thats exactly what I did. I'm not sure why the events were not getting
set (I set my layout correctly in config) but I just set my own using your
functions and it worked perfectly! thanks

**Comment by AdamBrodzinski** [comment]
Nice! :beers:  In the next release I think i'll experiment with bubbling to `<body>` instead if the layout isn't configured (I believe Blaze will let you do this now). I also recently ran into the use case of having multiple templates and this only accepts one... doh!

**Comment by tyler-dunkel** [comment]
I think I might have some time later this week to do the custom messages. Do you have an specifications for PRs (dev branch or whatever)?

**Comment by AdamBrodzinski** [comment]
Very cool! a dev branch would be ideal so I can add docs and bump the version (unless you get those too).

**Comment by tyler-dunkel** [comment]
alright I'll add to the docs and bump the version too. If I have extra time ill look into bubbling up to the body as well, if not i'll just add the custom messages and submit the PR. Would you like anything added to the logger about it? I assume just pointing it out in the docs is enough.

**Comment by sferoze** [comment]
@AdamBrodzinski Just wanted to say awesome package! Really simplified my code! I love it.

**Comment by AdamBrodzinski** [comment]
@sferoze Thanks!!

One thing you can do to clean it up even further is to abstract away the input markup with a blaze template... I did this when I had around 200 inputs to place! Those used defaults so that I only had to override the defaults by adding more props to the blaze template include.

``` html
{{> input label='First Name' name='fname' cols=4 required=1 }}
```

instead of the messy block of HTML

``` html
 <div class='fname__group col-md-4 form-group'>
    <label for='fname'>First Name</label>

    <input type='text' name='fname' 
      class='form-control'
      data-required data-onblur>

    <span class='err-msg'></span>
</div>
```
